docs/guides/tutorials/chatgpt_bot.rst

```diff
@@ -0,0 +1,2318 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For the embeddings request, we will call the ``openai.embeddings.create``
+method, in a new function called ``getEmbedding``.
+
+.. code-block:: typescript
+    :caption: app/api/generate-answer/route.ts
+
+    async function getEmbedding(query: string) {
+      const embeddingResponse = await openai.embeddings.create({
+        model: "text-embedding-ada-002",
+        input: query.replaceAll("\n", " "),
+      });
+
+      const [{ embedding }] = embeddingResponse.data;
+
+      return embedding;
+    }
+
+This new function again takes the question (as ``query``). We call the OpenAI
+library's ``embeddings.create`` method, specifying the model to use for
+generation (the ``model`` property of the options passed to the method) and
+passing the input (``query`` with all newlines replaced by single spaces).
+
+
+Get related documentation sections request
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let's look at the database query that will give us back the related sections in
+a variable named ``getSectionsQuery``.
+
+.. code-block:: typescript
+    :caption: app/api/generate-answer/route.ts
+
+    const getSectionsQuery = e.params(
+        {
+            target: e.OpenAIEmbedding,
+            matchThreshold: e.float64,
+            matchCount: e.int16,
+            minContentLength: e.int16,
+        },
+        (params) => {
+            return e.select(e.Section, (section) => {
+            const dist = e.ext.pgvector.cosine_distance(
+                section.embedding,
+                params.target
+            );
+            return {
+                content: true,
+                tokens: true,
+                dist,
+                filter: e.op(
+                    e.op(
+                      e.len(section.content),
+                      ">",
+                      params.minContentLength
+                    ),
+                    "and",
+                    e.op(dist, "<", params.matchThreshold)
+                ),
+                order_by: {
+                    expression: dist,
+                    empty: e.EMPTY_LAST,
+                },
+                limit: params.matchCount,
+            };
+            });
+        }
+    );
+
+In the above code, we use EdgeDB's TypeScript query builder to create a query.
+The query takes a few parameters:
+
+* ``target``: Embedding array to compare against to find related sections. In
+  this case, these will be the questions's embeddings we just generated.
+* ``matchThreshold``: Similarity threshold. Only matches with a similarity
+  score below this threshold will be returned. This will be a number between
+  ``0.0`` and ``1.0``. Values closer to ``0.0`` mean the documentation sections
+  must be very similar to the question while values closer to ``1.0`` allow for
+  more variance.
+* ``matchCount``: Maximum number of sections to return
+* ``minContentLength``: Minimum number of characters the sections should have
+  in order to be considered
```

Comment:
Is there a way to make list items more consistent in the terms of punctuation we use? If we list them, I think it's fine to separate them with `,`, if we have longer explanations and sentences to finish them with `.`.

Probably it's also fine to mix things, not many people pay attention to this.

Reply:
I'm OK making sure every list item ends with a period in a case where any of them do for a given list. I have a consistent rule I apply for this, but that doesn't always result in a consistent aesthetic within a given list. Maybe the better rule is, "if any items in a list end in a period, they all do." Would that work better? If an entire list is contains sentence fragments, it doesn't feel right to me to end those with a period.

If that does work better, this feels like something that should go in an issue that we can revisit later. I don't think it's critical to getting this published.